core/adaptive/adaptive_manager.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 1999-2020 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adaptive
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/alibaba/sentinel-golang/core/system_metric"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/pkg/errors"
+)
+
+// ControllerGenFunc represents the Controller generator function of a specific adaptive type.
+type ControllerGenFunc func(r *Config) (Controller, error)
+
+var (
+	acMap             = make(map[string]Controller, 0)
```

Comment:
if u init a map or array with 0 cap, u need not set it. Just delete it, pls.

Reply:
@AlexStocks Do you mean `var tcGenFuncMap map[trafficControllerGenKey]TrafficControllerGenFunc`?

Reply:
acMap             = make(map[string]Controller)